uarray/_uarray_dispatch.cxx

```diff
@@ -0,0 +1,491 @@
+	}
+
+	auto replaced_args = py_ref::steal(PySequence_Tuple(res));
+	if (!replaced_args)
+		return {};
+
+	auto replacer_args = py_make_tuple(args, kwargs, replaced_args);
+	if (!replacer_args)
+		return {};
+
+	res = py_ref::steal(PyObject_Call(replacer_, replacer_args, nullptr));
+	if (!res)
+		return {};
+
+	if (!PyTuple_Check(res) || PyTuple_Size(res) != 2)
+	{
+		PyErr_SetString(PyExc_TypeError,
+		                "Argument replacer must return a 2-tuple (args, kwargs)");
+		return {};
+	}
+
+	auto new_args = py_ref::ref(PyTuple_GET_ITEM(res.get(), 0));
+	auto new_kwargs = py_ref::ref(PyTuple_GET_ITEM(res.get(), 1));
+
+	new_kwargs = canonicalize_kwargs(new_kwargs);
+
+	if (!PyTuple_Check(new_args) || !PyDict_Check(new_kwargs))
+	{
+		PyErr_SetString(PyExc_ValueError, "Invalid return from argument_replacer");
+		return {};
+	}
+
+	return {std::move(new_args), std::move(new_kwargs)};
+}
+
+
+PyObject * Function_call(
+	Function * self, PyObject * args, PyObject * kwargs)
+{
+	return self->call(args, kwargs);
+}
+
+
+static py_ref BackendNotImplementedError;
+
+
+PyObject * Function::call(PyObject * args_, PyObject * kwargs_)
+{
+	auto args = canonicalize_args(args_);
+	auto kwargs = canonicalize_kwargs(kwargs_);
+
+	auto backends = py_ref::steal(
+		PyObject_Call(backends_, py_make_tuple(), nullptr));
+	if (!backends)
+		return nullptr;
+
+	auto i_backends = py_ref::steal(PyObject_GetIter(backends));
+	if (!i_backends)
+	{
+		PyErr_SetString(PyExc_TypeError, "backends must return an iterable");
+		return nullptr;
+	}
+
+	while (auto be = py_ref::steal(PyIter_Next(i_backends)))
+	{
+		auto backend = py_ref::steal(PyObject_GetAttrString(be, "backend"));
+		auto coerce = py_ref::steal(PyObject_GetAttrString(be, "coerce"));
+
+		if (!backend || !coerce)
+		{
+			PyErr_SetString(PyExc_TypeError, "Invalid backend spec");
+			return nullptr;
+		}
+
+		if (!PyBool_Check(coerce))
+		{
+			PyErr_SetString(PyExc_TypeError, "coerce must be a Bool");
+			return nullptr;
+		}
+
+		auto new_args = replace_dispatchables(backend, args, kwargs, coerce);
+		if (new_args.args == Py_NotImplemented)
+		{
+			continue;
+		}
+		if (new_args.args == nullptr)
+		{
+			return nullptr;
+		}
+
+		auto ua_function =
+			py_ref::steal(PyObject_GetAttrString(backend, "__ua_function__"));
+		if (!ua_function)
+		{
+			PyErr_SetString(PyExc_TypeError, "Backend must have __ua_function__");
+			return nullptr;
+		}
+
+		auto ua_func_args = py_make_tuple(
+			reinterpret_cast<PyObject *>(this), new_args.args, new_args.kwargs);
+		auto result = py_ref::steal(
+			PyObject_Call(ua_function, ua_func_args, nullptr));
+		if (result != Py_NotImplemented)
+			return result.release();
+	}
+
+	if (PyErr_Occurred())
+		return nullptr;
+
+	if (def_impl_ == Py_None)
+	{
+		PyErr_SetString(
+			BackendNotImplementedError,
+			"No selected backends had an implementation for this function.");
+		return nullptr;
+	}
+
+	return PyObject_Call(def_impl_, args, kwargs);
```

Comment:
I should probably draw attention to the fact that this is different from the current behaviour. The current behaviour requires calls to `with ua.set_backend` which would be difficult to do here (though not impossible).

Reply:
There was a reason I did that: Suppose the default implementation is in terms of other multi methods. That could be an issue.

Reply:
This is hard in general: We need this current backend to be the _only backend_ for the default implementation, while allowing it to rely on other backends. The only way I found is to mix a skip and a coerce. I can push a test that does this...

Reply:
Test in 604446ab904f5034c55dc85f10a2c2a6a5d1f155.